EIPS/eip-web3-access.md

```diff
@@ -0,0 +1,109 @@
+
+MetaMask and most other tools that provide access to web3-enabled environments do so automatically and without user consent. This exposes users of such environments to fingerprinting attacks since untrusted websites can check for a `web3` object and reliably identify web3-enabled clients.
+
+This proposal outlines a new dapp initialization strategy in which websites request access to the web3 API instead of relying on its preexistence in a given environment.
+
+## Specification
+
+### Typical dapp initialization
+
+```
+START dapp
+IF web3 is defined
+    CONTINUE dapp
+IF web3 is undefined
+    STOP dapp
+```
+
+### Proposed dapp initialization
+
+```
+START dapp
+IF web3 is defined
+    CONTINUE dapp
+IF web3 is undefined
+    REQUEST[1] web3 API
+    IF user approves
+        INJECT[2] web3 API
+        NOTIFY[3] dapp
+        CONTINUE dapp
+    IF user rejects
+        STOP dapp
+```
+
+**REQUEST[1]** This operation would be achieved by an implementation-level messaging API like `window.postMessage` and should pass a payload containing a `type` property with a value of “WEB3_API_REQUEST”.
+
+**INJECT[2]** This operation would be achieved by any implementation-level API that can expose the web3 API to the user’s browser context, such as HTML script tag injection.
+
+**NOTIFY[3]** This operation would be achieved by an implementation-level messaging API like `window.postMessage` and should pass a payload containing a `type` property with a value of “WEB3_API_SUCCESS” after successful web3 exposure or “WEB3_API_ERROR” after unsuccessful web3 exposure. In the case of an error, an optional `message` property can be included with additional information.
+
+### Example implementation: `postMessage`
+
+```js
+if (typeof web3 !== 'undefined') {
+    // web3 API defined, start dapp
+} else {
+    window.addEventListener('message', function (event) {
+        if (!event.data || !event.data.type) { return; }
+        if (event.data.type === 'WEB3_API_SUCCESS') {
+            // web3 API defined, start dapp
+        } else if (event.data.type === 'WEB3_API_ERROR') {
+            // Something went wrong, stop dapp
+        }
+    });
+    // request web3 API
+    window.postMessage({ type: 'WEB3_API_REQUEST' });
+}
+```
+
+## Rationale
+
+An [open issue](https://github.com/MetaMask/metamask-extension/issues/714) against the [MetaMask](https://github.com/MetaMask/metamask-extension) extension has received community upvotes and Richard Burton of the [Balance](https://github.com/balance-io) team published a well-received [blog post](https://medium.com/@ricburton/metamask-walletconnect-js-b47857efb4f7) discussing these potential changes.
+
+### Constraints
+
+* Web3 SHOULD NOT be exposed to websites by default.
+* Dapps SHOULD request web3 if it does not exist.
+* Users SHOULD be able to approve or reject web3 access.
+* Web3 SHOULD be exposed to websites after user consent.
+* Dapps MUST continue to work in environments that continue to auto-expose web3.
```

Comment:
I don't think there is value in asserting what dapps **MUST** do in this standard.  In fact, I think prescribing behavior of dapps isn't appropriate for this EIP outside of providing examples or workflows.  IMO, this standard should describe the API for web3 enabled browsers and dapp developers can then code against that however they want.